App/src/main/java/com/gluonhq/chat/service/WaveService.java

```diff
@@ -177,16 +180,51 @@ public void gotMessage(String senderUuid, String content, long timestamp) {
     public void gotMessage(String senderUuid, String content, long timestamp, String receiverUuid) {
         wave.getWaveLogger().log(Level.DEBUG,
                 "GOT MESSAGE from " + senderUuid + " for "+ receiverUuid+" with content " + content);
-        Channel dest = this.channels.stream()
-                .filter(c -> c.getMembers().size() > 0)
-                .filter(c -> c.getMembers().get(0).getId().equals(receiverUuid))
-                .findFirst().get();
+        Channel dest = getChannelByUuid(senderUuid);
         User sender = findUserByUuid(senderUuid, users).get();
-        ChatMessage chatMessage = new ChatMessage(content, sender, LocalDateTime.now());
+        ChatMessage chatMessage = new ChatMessage(content, sender, timestamp);
         Platform.runLater(() -> dest.getMessages().add(chatMessage));
         storeMessage(receiverUuid, senderUuid, content, timestamp);
     }
 
+    @Override
+    public void gotTypingAction(String senderUuid, boolean startTyping, boolean stopTyping) {
+        Channel dest = getChannelByUuid(senderUuid);
+        Platform.runLater(() -> {
+            if (startTyping) {
+                dest.typing().set(true);
+            }
+            if (stopTyping) {
+                dest.typing().set(false);
+            }
+        });
+    }
+
+    @Override
+    public void gotReceiptMessage(String senderUuid, int type, List<Long> timestamps) {
+        Channel dest = getChannelByUuid(senderUuid);
+        Platform.runLater(() -> {
+            FilteredList<ChatMessage> filtered = dest.getMessages()
+                    .filtered(m -> timestamps.contains(m.getTimestamp()));
+            ChatMessage.ReceiptType rtype = ChatMessage.ReceiptType.UNKNOWN;
+            switch (type) {
```

Comment:
Probably better to add a static method to `ChatMessage.ReceiptType`, like `valueOf(int)`